server/src/main/java/org/tctalent/server/service/db/impl/LinkPreviewServiceImpl.java

```diff
@@ -0,0 +1,289 @@
+      if (!ogTitle.isEmpty()) return ogTitle;
+    }
+
+    Elements twitterTitleElements = document.select("meta[name=\"twitter:title\"]");
+    if (!twitterTitleElements.isEmpty()) {
+      String twitterTitle = twitterTitleElements.first().attr("content");
+      if (!twitterTitle.isEmpty()) return twitterTitle;
+    }
+
+    Elements h1Elements = document.select("h1");
+    if (!h1Elements.isEmpty()) {
+      String h1 = h1Elements.first().text();
+      if (!h1.isEmpty()) return h1;
+    }
+
+    Elements h2Elements = document.select("h2");
+    if (!h2Elements.isEmpty()) {
+      String h2 = h2Elements.first().text();
+      if (!h2.isEmpty()) return h2;
+    }
+
+    return null;
+  }
+
+  private @Nullable String getDescription(Document document) {
+
+    Elements ogDescriptionElements = document.select("meta[property=\"og:description\"]");
+    if (!ogDescriptionElements.isEmpty()) {
+      String ogDescription = ogDescriptionElements.first().attr("content");
+      if (!ogDescription.isEmpty()) return ogDescription;
+    }
+
+    Elements twitterDescriptionElements = document.select("meta[name=\"twitter:description\"]");
+    if (!twitterDescriptionElements.isEmpty()) {
+      String twitterDescription = twitterDescriptionElements.first().attr("content");
+      if (!twitterDescription.isEmpty()) return twitterDescription;
+    }
+
+    Elements metaDescriptionElements = document.select("meta[name=\"description\"]");
+    if (!metaDescriptionElements.isEmpty()) {
+      String metaDescription = metaDescriptionElements.first().attr("content");
+      if (!metaDescription.isEmpty()) return metaDescription;
+    }
+
+    Elements paragraphs = document.select("p");
+    if (!paragraphs.isEmpty()) {
+      String firstPara = paragraphs.first().text();
+      if (!firstPara.isEmpty()) return firstPara;
+    }
+
+    return null;
+  }
+
+  private @Nullable String getDomain(Document document, String url) throws MalformedURLException {
+    Elements canonicalLinkElements = document.select("link[rel=canonical]");
+    if (!canonicalLinkElements.isEmpty()) {
+      String canonicalLink = canonicalLinkElements.first().attr("href");
+      if (!canonicalLink.isEmpty()) {
+        String domain = URI.create(canonicalLink).toURL().getHost().replace("www.", "");
+        if (domain != null && !domain.isEmpty()) return domain;
+      }
+    }
+
+    Elements ogUrlMetaElements = document.select("meta[property=\"og:url\"]");
+    if (!ogUrlMetaElements.isEmpty()) {
+      String ogUrlMeta = ogUrlMetaElements.first().attr("content");
+      if (!ogUrlMeta.isEmpty()) {
+        String domain = URI.create(ogUrlMeta).toURL().getHost().replace("www.", "");
+        if (domain != null && !domain.isEmpty()) return domain;
+      }
+    }
+
+    String domain = URI.create(url).toURL().getHost().replace("www.", "");
+    if (domain != null && !domain.isEmpty()) return domain;
+
+    return null;
+  }
+
+  private @Nullable String getImageUrl(Document document) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 208-225 into a separate method.